BanchoBot/functions/tournaments/matchup/runMatchup.ts

```diff
@@ -0,0 +1,328 @@
+
+        if (started || mpLobby.playing)
+            return;
+
+        if (
+            (
+                matchup.stage!.stageType === StageType.Qualifiers &&
+                matchup.teams!.some(team => !mpLobby.slots.some(m => m !== null && m.user.id.toString() === team.manager.osu.userID))
+            ) ||
+            (
+                matchup.stage!.stageType !== StageType.Qualifiers &&
+                !mpLobby.slots.some(m => m !== null && m.user.id.toString() === matchup.team1!.manager.osu.userID) &&
+                !mpLobby.slots.some(m => m !== null && m.user.id.toString() === matchup.team2!.manager.osu.userID)
+            )
+        ) {
+            await mpChannel.sendMessage(`Yo ${newPlayer.user.username} we're just waiting for all the ${matchup.stage!.tournament.matchupSize === 1 ? "players" : "managers"} to be in here and then we'll start the match`);
+            return;
+        }
+
+        started = true;
+        await mpChannel.sendMessage("OK WE;'RE STARTING THE MATCH let's go");
+
+        try {
+            log(matchup, "Picking map");
+            await loadNextBeatmap(matchup, mpLobby, mpChannel, pools, autoStart, false);
+            log(matchup, `Map picked: ${mpLobby.beatmapId} with mods ${mpLobby.mods.map(mod => mod.shortMod).join(", ")}`);
+        } catch (ex) {
+            await mpChannel.sendMessage(`Error loading beatmap: ${ex}`);
+            log(matchup, `Error loading beatmap: ${ex}`);
+        }
+    });
+
+    // Player left event
+    mpLobby.on("playerLeft", async (player) => {
+        log(matchup, `Player ${player.user.username} left the lobby`);
+
+        if (
+            mpLobby.playing &&
+            playersInLobby.some(p => p.user.id === player.user.id)
+        ) {
+            const team = matchup.teams!.find(team => team.members.some(m => m.osu.userID === player.user.id.toString()));
+            if (
+                (team &&
+                    (
+                        aborts.get(team.ID) === undefined ||
+                        !matchup.stage!.tournament.teamAbortLimit ||
+                        aborts.get(team.ID)! <= matchup.stage!.tournament.teamAbortLimit
+                    )
+                ) && (
+                    matchStart &&
+                    Date.now() - matchStart.getTime() < (matchup.stage!.tournament.abortThreshold || 30) * 1000
+                )
+            ) {
+                await Promise.all([
+                    mpLobby.abortMatch(),
+                    mpChannel.sendMessage(`${player.user.username} left the match so we're aborting`),
+                ]);
+                aborts.set(team.ID, (aborts.get(team.ID) || 0) + 1);
+            }
+        }
+
+        playersInLobby = playersInLobby.filter(p => p.user.id !== player.user.id);
+    });
+
+    // All players ready event
+    mpLobby.on("allPlayersReady", async () => {
+        if (mapsPlayed.some(m => m.beatmap!.ID === mpLobby.beatmapId))
```

Comment:
i recommend running `await mpLobby.updateSettings()` after everyone is ready, especially if you're going to check mods, otherwise the data you're checking against will be incorrect as Bancho doesn't report mods updates.